libwasmvm/src/error/go.rs

```diff
@@ -15,77 +15,185 @@ use crate::memory::UnmanagedVector;
+            GoError::User => Err(BackendError::user_err(build_error_msg())),
+            // Everything else goes into unknown
+            GoError::CannotSerialize | GoError::Other => {
+                Err(BackendError::unknown(build_error_msg()))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use cosmwasm_vm::BackendError;
+
+    use super::{GoError, UnmanagedVector};
+
+    #[test]
+    fn go_error_into_result_works() {
+        let default = || "Something went wrong but we don't know".to_string();
+
+        let error = GoError::None;
+        let error_msg = UnmanagedVector::new(None);
+        let a = unsafe { error.into_result(error_msg, default) };
+        assert!(a.is_ok());
+
+        let error = GoError::Panic;
+        let error_msg = UnmanagedVector::new(None);
+        let a = unsafe { error.into_result(error_msg, default) };
+        assert!(matches!(a.unwrap_err(), BackendError::ForeignPanic {}));
+
+        let error = GoError::BadArgument;
+        let error_msg = UnmanagedVector::new(None);
+        let a = unsafe { error.into_result(error_msg, default) };
+        assert!(matches!(a.unwrap_err(), BackendError::BadArgument {}));
+
+        let error = GoError::OutOfGas;
+        let error_msg = UnmanagedVector::new(None);
+        let a = unsafe { error.into_result(error_msg, default) };
+        assert!(matches!(a.unwrap_err(), BackendError::OutOfGas {}));
+
+        // CannotSerialize maps to Unknown
+        let error = GoError::CannotSerialize;
+        let error_msg = UnmanagedVector::new(None);
+        let a = unsafe { error.into_result(error_msg, default) };
+        // TODO: use equality check on BackendError
+        match a.unwrap_err() {
+            BackendError::Unknown { msg } => assert_eq!(msg.unwrap(), default()),
+            err => panic!("Unexpected error: {}", err),
+        }
+
+        // GoError::User with none message
+        let error = GoError::User;
+        let error_msg = UnmanagedVector::new(None);
+        let a = unsafe { error.into_result(error_msg, default) };
+        // TODO: use equality check on BackendError
+        match a.unwrap_err() {
+            BackendError::UserErr { msg } => assert_eq!(msg, default()),
+            err => panic!("Unexpected error: {}", err),
+        }
+
+        // GoError::User with some message
+        let error = GoError::User;
+        let error_msg = UnmanagedVector::new(Some(Vec::from(b"kaputt" as &[u8])));
+        let a = unsafe { error.into_result(error_msg, default) };
+        // TODO: use equality check on BackendError
+        match a.unwrap_err() {
+            BackendError::UserErr { msg } => assert_eq!(msg, "kaputt"),
+            err => panic!("Unexpected error: {}", err),
+        }
+
+        // GoError::User with some message too long message
+        let error = GoError::User;
+        let error_msg = UnmanagedVector::new(Some(vec![0x61; 10000])); // 10000 times "a"
```

Comment:
interesting to check with some that uses utf8 code points... like some 3 bytes utf-8 char repeated. and ensure we get invalid utf8 error there. (or do we want "error too long" message?)

Reply:
We convert bytes to String using `String::from_utf8_lossy`. So if we break some UTF-8, it gives us a � and we don't have to handle UTF-8 errors.